mix_pki/json.go

```diff
@@ -0,0 +1,173 @@
+
+// Package provides mix PKI client implementations
+package mix_pki
+
+import (
+	"encoding/base64"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"strings"
+
+	"github.com/katzenpost/core/crypto/ecdh"
+	"github.com/katzenpost/core/pki"
+	"github.com/katzenpost/core/sphinx/constants"
+	"github.com/op/go-logging"
+)
+
+var log = logging.MustGetLogger("mixclient")
+
+type JsonMixDescriptor struct {
+	Name            string
+	ID              string
+	IsProvider      bool
+	LoadWeight      int
+	TopologyLayer   int
+	EpochAPublicKey string
+	EpochBPublicKey string
+	EpochCPublicKey string
+	Ipv4Address     string
+	TcpPort         int
+}
+
+type JsonStaticPKI struct {
+	MixDescriptors []JsonMixDescriptor
+}
+
+func JsonFromDescriptor(m *pki.MixDescriptor) *JsonMixDescriptor {
+	desc := JsonMixDescriptor{
+		Name:            m.Name,
+		ID:              base64.StdEncoding.EncodeToString(m.ID[:]),
+		IsProvider:      m.IsProvider,
+		LoadWeight:      int(m.LoadWeight),
+		TopologyLayer:   int(m.TopologyLayer),
+		EpochAPublicKey: base64.StdEncoding.EncodeToString(m.EpochAPublicKey.Bytes()),
+		EpochBPublicKey: base64.StdEncoding.EncodeToString(m.EpochBPublicKey.Bytes()),
+		EpochCPublicKey: base64.StdEncoding.EncodeToString(m.EpochCPublicKey.Bytes()),
+		Ipv4Address:     m.Ipv4Address,
+		TcpPort:         m.TcpPort,
+	}
+	return &desc
+}
+
+func (j *JsonMixDescriptor) MixDescriptor() (*pki.MixDescriptor, error) {
+	idBytes, err := base64.StdEncoding.DecodeString(j.ID)
+	if err != nil {
+		return nil, err
+	}
+	var id [constants.NodeIDLength]byte
+	copy(id[:], idBytes)
+	aBytes, err := base64.StdEncoding.DecodeString(j.EpochAPublicKey)
+	if err != nil {
+		return nil, err
+	}
+	keyA := ecdh.PublicKey{}
+	keyA.FromBytes(aBytes)
+	d := pki.MixDescriptor{
+		Name:            strings.ToLower(j.Name),
+		ID:              id,
+		IsProvider:      j.IsProvider,
+		LoadWeight:      uint8(j.LoadWeight),
+		TopologyLayer:   uint8(j.TopologyLayer),
+		EpochAPublicKey: &keyA,
+		Ipv4Address:     j.Ipv4Address,
+		TcpPort:         j.TcpPort,
+	}
+	return &d, nil
+}
+
+type StaticPKI struct {
+	MixMap      map[[constants.NodeIDLength]byte]*pki.MixDescriptor
+	ProviderMap map[string]*pki.MixDescriptor
+}
+
+func NewStaticPKI() *StaticPKI {
+	staticPKI := StaticPKI{
+		MixMap:      make(map[[constants.NodeIDLength]byte]*pki.MixDescriptor),
+		ProviderMap: make(map[string]*pki.MixDescriptor),
+	}
+	return &staticPKI
+}
+
+func (t *StaticPKI) GetDescriptor(id [constants.NodeIDLength]byte) (*pki.MixDescriptor, error) {
+	mix, ok := t.MixMap[id]
+	if ok {
+		return mix, nil
+	}
+	return nil, errors.New("mix id not found in static consensus")
+}
+
+func (t *StaticPKI) GetMixesInLayer(layer uint8) []*pki.MixDescriptor {
+	l := []*pki.MixDescriptor{}
+	for _, v := range t.MixMap {
+		if v.TopologyLayer == layer {
+			l = append(l, v)
+		}
+	}
+	return l
+}
+
+func (t *StaticPKI) GetLatestConsensusMap() *map[[constants.NodeIDLength]byte]*pki.MixDescriptor {
```

Comment:
Unneeded? a caller could access `.MixMap` directly

Reply:
not currently used in path selection